### DISCUSSION TITLE
Custom Pricing for logged in Users

### DISCUSSION THREAD

**Discussion by tomsvogel** [vendurehq/vendure#899] [Q&A] [answered]
Hey 👋 ,

I am trying estimate a solution for next/commerce + vendure but i have an special use case.

Is there a chance to create a custom pricing method for logged in users.

An how the current logged in User is transfered to the api?

thanks for any hints in advance!

best regards toms

**Comment by michaelbromley** [comment, marked as answer]
Hi there!

I think you should be able to use the [OrderItemPriceCalculationStrategy](https://www.vendure.io/docs/typescript-api/orders/order-item-price-calculation-strategy/) to implement this requirement. In that function, you have access to the `RequestContext` which you can use to see if the current session is with an logged-in user.

> An how the current logged in User is transfered to the api?

Sessions are managed via cookies or bearer tokens as described here: https://www.vendure.io/docs/storefront/managing-sessions/

**Reply by tomsvogel** [reply]
oh thanks that was fast, i will try!

**Reply by tomsvogel** [reply]
@michaelbromley could you please give me an other hint, how to get the usergroups of the current signed in user in the OrderItemPriceCalculationStrategy. 🙏

**Reply by michaelbromley** [reply]
Sure! So Vendure uses NestJS and as a result you can take advantage of Nest's dependency injection system to inject other services into your OrderItemPriceCalculationStrategy. This is done by passing the required service into the constructor as a private field. In this case I don't think there's a built-in service method which returns CustomerGroups for a given userId, so you'd need to create a DB query yourself to do this. Example (not tested, you might need to tweak):

```ts
class MyOrderItemPriceCalculationStrategy implements OrderItemPriceCalculationStrategy {

  constructor(private connection: TransactionalConnection, private customerService: CustomerService) {}
  
  async calculateUnitPrice(ctx: RequestContext) {
    const customer = await this.customerService.findOneByUserId(ctx, ctx.activeUserId);
    const groups = await this.connection.getRepository(ctx, CustomerGroup)
      .createQueryBuilder('group')
      .leftJoin('goup.customers', 'customer')
      .where('customer.id = :customerId', { customerId: customer.id })
      .getMany();
  }

}
```

**Reply by tomsvogel** [reply]
could you recommend me a good learning resource for it, never worked with nest.js 

the config now wants me to add the connection and customerService... 

![image](https://user-images.githubusercontent.com/2229728/120211241-fe8ffc80-c230-11eb-8d8e-589e657ad6dd.png)

**Reply by tomsvogel** [reply]
and thanks for your quick responses!!!

**Reply by michaelbromley** [reply]
Oh wait, I told you wrong, sorry. You actually need to use add an `init` method to your strategy, as described here: https://www.vendure.io/docs/typescript-api/common/injectable-strategy/#injectablestrategy

So it would actually look like:

```ts
class MyOrderItemPriceCalculationStrategy implements OrderItemPriceCalculationStrategy {
  private connection: TransactionalConnection;
  private customerService: CustomerService;

  init(injector: Injector) {
    this.connection = injector.get(TransactionalConnection);
    this.customerService = injector.get(CustomerService);
  }
  
  async calculateUnitPrice(ctx: RequestContext) {
    // .. as above
  }

}
```

The usual pattern for dependency injection in NestJS  _is_ done by injecting into a class constructor, as described here: https://docs.nestjs.com/fundamentals/custom-providers#di-fundamentals but in the case of Vendure strategies, we cannot do it like this for various good reasons, which is why we provide the `init()` method as a hook to allow injection of dependencies.
